### ISSUE TITLE
Refetch on window focus

### ISSUE THREAD

**Issue by tylerzey** [apollographql/apollo-feature-requests#422]
Libraries such as SWR and React Query support opting into queries re-fetching upon the window being re-focused.

They allow to you configure this behavior on a per-query or on a global level.  It seems quite popular and might be a feature Apollo would like to support.

Example from [React Queries docs](https://tanstack.com/query/v4/docs/framework/react/guides/window-focus-refetching)
```
useQuery({
  refetchOnWindowFocus: false,
})
```

A previous issue (feature request) had a lot of comments on it [asking for Apollo to support the feature](https://github.com/apollographql/apollo-feature-requests/issues/247).

To implement this feature in Apollo current, you have to do the following.

On a global level:

```
const apolloClient = new ApolloClient();

window.addEventListener('focus', () => {
  apolloClient.refetchQueries({ include: 'active' });
});
```

On a per-query level:

https://github.com/apollographql/apollo-feature-requests/issues/247#issuecomment-1063411159

Specifically: 

```
  const { isWindowFocused } = useWindowFocus();

  useEffect(() => {
    if (!isWindowFocused) {
      stopPolling();
    } else {
      // Refetch data immediately when the window is refocused.
      refetch?.();
      startPolling(pollInterval);
    }
  }, [isWindowFocused, pollInterval, refetch, startPolling, stopPolling]);
```

The use case seems fairly common and might be something to consider bringing into Apollo as a feature.

**Comment by nwazuo** [comment]
+1 for this

**Comment by shyamayadav154** [comment]
+1

**Comment by optimistic-updt** [comment]
+1

**Comment by rajat-sr** [comment]
+1

**Comment by stanislavyv** [comment]
+1

**Comment by man-trackunit** [comment]
+2

**Comment by KMKoushik** [comment]
This is how I do it right now

```js
useEffect(() => {
    const refetchQueries = () => {
      if (document.visibilityState === "visible") {
        client.refetchQueries({ include: "active" });
      }
    };

    window.addEventListener("visibilitychange", refetchQueries);

    return () => {
      window.removeEventListener("visibilitychange", refetchQueries);
    };
  }, []);
```

**Comment by nextglabs** [comment]
+1

**Comment by diegomargeniss** [comment]
+1

**Comment by felps-dev** [comment]
+1

**Comment by nachoargentina** [comment]
+1

**Comment by ladderschool** [comment]
+1

**Comment by rlittle-wavefin** [comment]
+1

**Comment by matheusAle** [comment]
+1

**Comment by christiancuri** [comment]
+1

**Comment by jerelmiller** [comment]
Hey all 👋 

I've got some exciting news to share! We will be including this feature in the upcoming release of 4.2.0 🎉. We just put out an alpha version that includes support for window focus refetching that you can try out today. It looks like this:

```ts
import { RefetchEventManager, windowFocusSource } from "@apollo/client";

const client = new ApolloClient({
  // ...
  refetchEventManager: new RefetchEventManager({
    sources: {
      windowFocus: windowFocusSource
    }
  })
});

// disable automatic refetch per-query
useQuery(query, { refetchOn: { windowFocus: false } });
```

Please see the [release notes](https://github.com/apollographql/apollo-client/releases/tag/%40apollo%2Fclient%404.2.0-alpha.4) for a more comprehensive overview of how this feature works. Install it using the `next` tag:

```ts
npm install @apollo/client@next
```

I'd love any and all feedback you have! While the feature is in alpha, there is time to address change if something doesn't work or feel right, so please try it out and let us know how you like it.

Thanks all for your patience! I'm looking forward to getting a highly requested feature in place!

**Comment by mogelbrod** [comment]
@jerelmiller I haven't used the new API yet but would love for a straight forward way to trigger a refetch on triggers like:
1. The window (or React Native app) regains focus after being unfocused for over X seconds
2. The GraphQL WebSocket connection gets disconnected for more than X seconds

If the API doesn't already support cases like these yet, perhaps it could be tweaked/generalized to open up for use cases like the ones above?

**Comment by jerelmiller** [comment]
@mogelbrod yep it should everything you're looking for! Its designed with flexibility in mind so that you can add your own sources that trigger refetches from whatever events make sense. That looks like this:

```ts
const refetchEventManager = new RefetchEventManager({
  sources: {
    yourOwnCustomSource: () => {
      return new Observable((observer) => {
        function handleEvent() {
          // call observer.next to trigger the refetch
          observer.next();
        }

        window.addEventListener("someEvent", handleEvent);

        return () => window.removeEventListener("someEvent", handleEvent)
      });
    },
    anotherSource: () => {
      // you can also use built-in RxJS constructs
      return fromEvent(window, "someEvent");
    }
  }
});
```

You can also emit those events manually to trigger refetches:

```ts
refetchEventManager.emit("yourOwnCustomSource");
```

Of course your source doesn't have to use `window.addEventListener` like this example and can be anything you want, but gives you an idea of what this looks like.

If you're curious what the full API looks like, check out https://github.com/apollographql/apollo-client/pull/13228 which are the docs for this feature. I'm waiting for a docs preview link that I can share, but for now this should give you an idea of what its capable of.

We are planning to move to release candidate phase today with a public release very soon. I'd love if you could try it out before next week in case there is something that should be addressed in the design of this feature! Once its publicly released, we won't be able to change the API in non-breaking ways.

**Comment by mogelbrod** [comment]
Lovely, that should cover it! Thanks for the quick reply 😃

**Comment by jerelmiller** [comment]
Hey all 👋 

We just released version [4.2](https://github.com/apollographql/apollo-client/releases/tag/%40apollo%2Fclient%404.2.0) which includes this feature. Please see the [Event-based refetching guide](https://www.apollographql.com/docs/react/data/event-based-refetching) to learn more about using this in your applications. Thanks again for your patience on this feature!
